webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/JsonSchema.scala

```diff
@@ -0,0 +1,598 @@
+  }
+  def forbiddenError(schema: ResponseSchema, errorMsg: String)(implicit prettify: Boolean) = {
+    generic.forbiddenError(JsonRudderApiResponse.error(schema, errorMsg))
+  }
+
+  //import that to transform a class from JsonResponse into a lift response. An encoder for the JsonResponse classe
+  // must be available.
+  trait implicits {
+    implicit class ToLiftResponseList[A](result: IOResult[Seq[A]]) {
+      def toLiftResponseList(params: DefaultParams, schema: ResponseSchema)(implicit encoder: JsonEncoder[A]): LiftResponse = {
+        implicit val prettify = params.prettify
+        result.fold(
+          err => internalError(schema, err.fullMsg)
+        , seq => successList(schema, seq.toList)
+        ).runNow
+      }
+      def toLiftResponseList(params: DefaultParams, schema: EndpointSchema)(implicit encoder: JsonEncoder[A]): LiftResponse = {
+        toLiftResponseList(params, ResponseSchema.fromSchema(schema))
+      }
+    }
+    implicit class ToLiftResponseOne[A](result: IOResult[A]) {
+      def toLiftResponseOne(params: DefaultParams, schema: ResponseSchema, id: A => String)(implicit encoder: JsonEncoder[A]): LiftResponse = {
+        implicit val prettify = params.prettify
+        result.fold(
+          err => internalError(schema, err.fullMsg)
+        , one => successOne(schema, one, id(one))
+        ).runNow
+      }
+      def toLiftResponseOne(params: DefaultParams, schema: EndpointSchema, id: A => String)(implicit encoder: JsonEncoder[A]): LiftResponse = {
+        toLiftResponseOne(params, ResponseSchema.fromSchema(schema), id)
+      }
+      // when the computation give the response schema
+      def toLiftResponseOneMap[B](params: DefaultParams, errorSchema: ResponseSchema, map: A => (ResponseSchema, B, String))(implicit encoder: JsonEncoder[B]): LiftResponse = {
+        implicit val prettify = params.prettify
+        result.fold(
+          err => internalError(errorSchema, err.fullMsg)
+        , one => {
+            val (schema, x, id) = map(one)
+            successOne(schema, x, id)
+          }).runNow
+      }
+    }
+
+  }
+}
+
+//////////////////////////// Case classes used in JSON answers ////////////////////////////
+
+// to avoid ambiguity with corresponding business objects, we use "JR" as a prfix
+object JsonResponseObjects {
+  final case class JRRule(
+      changeRequestId : Option[String] = None
+    , id              : String
+    , displayName     : String
+    , categoryId      : String
+    , shortDescription: String
+    , longDescription : String
+    , directives      : List[String] //directives ids
+    , targets         : List[JRRuleTarget]
+    , enabled         : Boolean
+    , system          : Boolean
+    , tags            : List[Map[String, String]]
+  )
+  object JRRule {
+    // create an empty json rule with just ID set
+    def empty(id: String) = JRRule(None, id, "", "", "", "", Nil, Nil, false, false, Nil)
```

Comment:
why do we need an empty one ?

Reply:
It's in the case of deletion, when the rule is absent, we don't have more info, only the id.